src/data/nat/basic.lean

```diff
@@ -278,6 +281,17 @@ iff.intro eq_zero_of_mul_eq_zero (by simp [or_imp_distrib] {contextual := tt})
 @[simp] protected theorem zero_eq_mul {a b : ℕ} : 0 = a * b ↔ a = 0 ∨ b = 0 :=
 by rw [eq_comm, nat.mul_eq_zero]
 
+lemma eq_zero_of_double_le {a : ℕ} (h : 2 * a ≤ a) : a = 0 :=
```

Comment:
Wouldn't it make sense to generalise `2` to an arbitrary nat ≥ 2?

Reply:
I guess so. There is a trade-off. Until `library_search` is cleverer about deciding to use `dec_trivial` and `norm_num` to fill in some arguments, it's harder to find more general lemmas that have arithmetic side conditions. This is the lemma I wanted, so I just wrote this one.

Reply:
Oh well, we have lots of special lemmas like `two_ne_zero`. So I guess it's fine to have this special case.

Reply:
I agree this lemma is a bit specialized. (I wouldn't say two_ne_zero is specialized in the same way.) In particular I agree with Johan that this should be proven in generality, even if you later specialize it to the case 2. In the case 2, I think this generalizes to any cancellative ordered semiring (with conclusion a <= 0, as an iff). The case n doesn't generalize quite as well unless a + a + ... + a <= 0 implies a <= 0 - I guess that holds in linear ordered group structures.

Reply:
FYI, one way to condense this (without Mario's full generalization):
```lean
lemma eq_zero_of_double_le {a : ℕ} (h : 2 * a ≤ a) : a = 0 :=
nat.eq_zero_of_le_zero $ 
  by rwa [two_mul, nat.add_le_to_le_sub, nat.sub_self] at h; refl

lemma eq_zero_of_mul_le {a b : ℕ} (hb : 2 ≤ b) (h : b * a ≤ a) : a = 0 :=
eq_zero_of_double_le $ le_trans (nat.mul_le_mul_right _ hb) h
```

I hope most of the proofs here can be condensed similarly.

Reply:
Thanks, @robertylewis, I've included your proofs.